### PR TITLE
Parsing: Use full parser in `do_blocks` with nested block support

### PR DESCRIPTION
**Pull request by dmsnell**
Updates `do_blocks()` and `gutenberg_render_block()` so that we can support nested blocks inside of dynamic blocks. This replaces the use of the partial parser which extracts registered dynamic blocks with the full parser.

This change will allow dynamic blocks which contain nested blocks inside of them and it will pave the way for a filtering API to structurally process blocks.

The partial parser came about at a time before the default parser was written; it was faster than the spec parser and was a tradeoff to get dynamic blocks rendering. The default parser, however, has been fast enough for a while to run on page render and so this PR exists to finally get it into the pipeline.

### Iteration

<del>The `RecursiveIteratorIterator` should provide us a way of performing the depth-first block traversal without risking a stack overflow. We keep rendered blocks on a stack and pop them off as we go.</del>

Interestingly enough I did some testing and `RecursiveIteratorIterator` was a let-down. There's one important thing it does which is bypass the `xdebug.max_nesting_level` which limits recursion depth when `xdebug` is running; sadly it's much much slower than basic recursion.

With the `RecursiveIteratorIterator` we can continue to nest into blocks until we run out of memory but the runtime performance is exponential. With recursion we hit the same memory limit if `xdebug` is disabled but hit it usually around 100 if it is. That means that for massive posts we could run into unexpected failures when debugging if we go the recursion route.

What is the memory limit? In my testing back to PHP 5.6 on my local laptop it was at around 100,000 levels of nesting - this should be obviously sufficient - I can't imagine an editor showing that many levels of nesting. With the recursive version and a limit of 100 that's still almost equally viable as 100,000. I'd recommend we stick with this recursive approach then unless we get bug reports of failures when debugging, at which point we can discuss bringing back the `RecursiveIteratorIterator` version.

On the other hand I may simply have been writing the `WP_Block_Tree_Iterator` in a stupid way that was way too slow. It would be worth running an experiment where I make a custom tree iterator without using the SPL functions and see if the performance characteristics are similar.

**Update**

After running some tests I made ended up with [a basic tree iterator](https://repl.it/@dmsnell/SickGreedyShockwave) and it hit hard performance issues long before the recursive version. I think that the memory churning of the stacks required to track nodes in the tree is far less efficient than using the built-in argument passing in functions and closures provided by PHP with the recursive method.

In other words, I am currently unable to come close to the recursive version in terms of performance. Does this matter with small nesting? Probably not. Should we then prefer the safe version that won't break with low `max_nesting_depth`? I don't know. Probably not since anything over 100 levels of nested blocks is somewhat of an insane construction.

### Testing and Review
The patch needs testing, but this can be hard because we don't have too many deeply-nested blocks. It's best to try with things like columns.

👉 I would greatly appreciate some help adding the appropriate deprecation messages for `get_dynamic_blocks_regex()` and friends!

### CODE REVIEW

File: lib/blocks.php [hunk 1]
```diff
@@ -164,11 +164,32 @@ function gutenberg_render_block( $block ) {
 		}
```
Comment (georgestephanis) [inline lib/blocks.php]: Line 163 above will also need to be adjusted to pass the content along -- as at the moment it doesn't.

Comment (georgestephanis) [reply lib/blocks.php]: I think ideally the function should be revised to something like --

```php
/**
 * Renders a single block into a HTML string.
 *
 * @since 1.9.0
 *
 * @param  array $block A single parsed block object.
 * @param  array $inner_blocks A list of processed inner blocks.
 * @return string String of rendered HTML.
 */
function gutenberg_render_block( $block, $inner_blocks = [] ) {
	$block_name = isset( $block['blockName'] ) ? $block['blockName'] : null;
	$block_type = $block_name ? WP_Block_Type_Registry::get_instance()->get_registered( $block_name ) : null;
	$attributes = is_array( $block['attrs'] ) ? $block['attrs'] : array();

	// If there are innerBlocks, stitch their results back in.
	if ( ! empty( $block['blockMarkers' ] ) ) {
		$output = '';
		$index = 0;
		foreach( $block['blockMarkers'] as $i => $p ) {
			$output .= substr( $block['innerHTML'], $index, $p - $index );
			$output .= $inner_blocks[ $i ];
			$index = $p;
		}
		$output .= substr( $block['innerHTML'], $index );

		$block['innerHTML'] = $output;
	}

	// Handle both dynamic and static.
	if ( null !== $block_type && $block_type->is_dynamic() ) {
		return $block_type->render( $attributes, $block['innerHTML'] );
	}
	return $block['innerHTML'];
}
```

**Comment by georgestephanis** [comment]
encapsulated ^^ into a pr, with unit tests:

https://github.com/WordPress/gutenberg/pull/11227

**Comment by youknowriad** [comment]
Moving out of 4.2 as we have to release soon.

**Comment by dmsnell** [comment]
Previous changes on this branch before the rebase…
https://github.com/WordPress/gutenberg/commit/4a4feba697858b8dc8043df1bcd422d11dc71410

**Comment by dmsnell** [comment]
@aduth if you look at the [failing tests](https://travis-ci.org/WordPress/gutenberg/jobs/452226181) you can see that we're hitting trailing-newline issues because in the existing `do_blocks()` we strip `[\r\n]` _after_ a block's closing comments but here we're not.

can you help me understand what our desired end-product is so I can implement it? or, if it doesn't matter, should I update the test fixtures?

if we look at the parser output we can see that these newlines end up as freeform blocks whose contents are exactly `\n\n`. from a WordPress standpoint I guess that could risk introducing `autop` paragraphs? from an HTML standpoint it's just extra meaningless whitespace.

**Comment by dmsnell** [comment]
rebased and squashed, previous head was 6d713cd

**Comment by aduth** [comment]
>if it doesn't matter

It doesn't seem like something which should matter. I think previously in stripping comments I'd anticipated the empty line where the comment had been in considering to do the trim.

**Comment by aduth** [comment]
>if we look at the parser output we can see that these newlines end up as freeform blocks whose contents are exactly \n\n. from a WordPress standpoint I guess that could risk introducing autop paragraphs? from an HTML standpoint it's just extra meaningless whitespace.

`autop` is disabled for posts which contian blocks.

I think the freeform newlines would be expected, or at least this also occurs in the editor during its block parse?

File: lib/blocks.php [hunk 2]
```diff
@@ -148,119 +148,61 @@ function get_dynamic_blocks_regex() {
...
-			$offset          = $block_match[0][1];
-			$block_name      = $block_match[1][0];
-			$is_self_closing = isset( $block_match[4] );
-
-			// Reset attributes JSON to prevent scope bleed from last iteration.
-			$block_attributes_json = null;
-			if ( isset( $block_match[3] ) ) {
-				$block_attributes_json = $block_match[3][0];
-			}
+		$without_trailing_newlines = preg_replace( '/(<!--\s+\/?wp:.*?-->)\r?\n?/m', '$1', $content );
```
Comment (aduth) [inline lib/blocks.php]: With prior comments in mind, I'm wondering: Do we really need this? Can we feed `$content` directly to `gutenberg_parse_blocks`?

Comment (dmsnell) [reply lib/blocks.php]: there's no technical ned to strip the newlines - mainly I was trying to preserve existing behavior but I don't see why they are needed. if we skip it altogether then the output parse will change but probably not in a semantically-altering way

of note: the existing system only strips the trailing newlines from blocks which aren't registered as dynamic blocks at the time of render, thus making it pretty confusing to match behaviors.

are you in favor of removing all the newline-stripping and update the tests instead?

Comment (aduth) [reply lib/blocks.php]: >are you in favor of removing all the newline-stripping and update the tests instead?

This sounds fine, yes.

**Comment by dmsnell** [comment]
sorry @aduth I really appreciated your review on this but I had to change something out so that dynamic blocks get their full rendered inner content.

this wasn't _necessary_ I suppose given the constraint that dynamic blocks have no inner blocks but at the same time it wasn't necessary for us to impose that and I know we'll be wanting this shorty.

if we provide the inner content here then it sets us up for success as we continue to improve the process I guess. your thoughts are welcome

**Comment by georgestephanis** [comment]
With that change, this unit test should pass if it's worth adding in here:

https://github.com/WordPress/gutenberg/pull/11227/files#diff-75e130118746d8abf068e5a2a090d145

File: phpunit/class-dynamic-blocks-render-test.php [hunk 3]
```diff
@@ -164,4 +168,69 @@ function test_dynamic_block_renders_string() {
...
 		$this->assertInternalType( 'string', $rendered );
 	}
+
+	function test_dynamic_block_gets_inner_html() {
+		register_block_type(
+			'core/dynamic',
+			array(
+				'render_callback' => array(
+					$this,
+					'render_serialize_dynamic_block'
```
Comment (gwwar) [inline phpunit/class-dynamic-blocks-render-test.php]: Tests are failing from the linter, we always want trailing commas here.

```
FILE: /app/phpunit/class-dynamic-blocks-render-test.php
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 5 LINES
----------------------------------------------------------------------
 178 | ERROR | [x] Each array item in a multi-line array declaration
     |       |     must end in a comma
     |       |     (WordPress.Arrays.CommaAfterArrayItem.NoComma)
 196 | ERROR | [x] Each array item in a multi-line array declaration
     |       |     must end in a comma
     |       |     (WordPress.Arrays.CommaAfterArrayItem.NoComma)
 205 | ERROR | [x] Each array item in a multi-line array declaration
     |       |     must end in a comma
     |       |     (WordPress.Arrays.CommaAfterArrayItem.NoComma)
 223 | ERROR | [x] Each array item in a multi-line array declaration
     |       |     must end in a comma
     |       |     (WordPress.Arrays.CommaAfterArrayItem.NoComma)
 229 | ERROR | [x] Whitespace found at end of line
     |       |     (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

File: lib/blocks.php [hunk 4]
```diff
@@ -148,119 +148,55 @@ function get_dynamic_blocks_regex() {
...
-		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
-		if ( null !== $block_type && $block_type->is_dynamic() ) {
-			return $block_type->render( $attributes );
-		}
+	$block_type    = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$is_dynamic    = $block['blockName'] && null !== $block_type && $block_type->is_dynamic();
+	$inner_content = '';
+	$index         = 0;
+
+	foreach ( $block['innerContent'] as $chunk ) {
```
Comment (gwwar) [inline lib/blocks.php]: Also a number of `Undefined index:  innerContent` errors on the 7.1 test suite.

Try guarding with:
```php
isset( $block['innerContent'] ) 
```

**Comment by gwwar** [comment]
Besides getting tests green is there anything else blocking, or areas that need more polish before landing?

File: lib/blocks.php (hunk 4 above)
Comment (dmsnell) [reply lib/blocks.php]: this is strange. if `innerContent` is unavailable then everything else should be broken. I'm stunned; will look into this

**Comment by pento** [comment]
`get_dynamic_blocks_regex()` hasn't been merged to Core, but it's still being used by `strip_dynamic_blocks()`.

Upon reflection, I don't think `strip_dynamic_blocks()` was the right solution for the problem (#7268) that it was originally intended to solve. Something more like what @chrisvanpatten [described](https://github.com/WordPress/gutenberg/issues/7268#issuecomment-400376787) would probably give us results more in line with how `wp_trim_excerpt()` is intended to behave.

I'll follow up with a PR after this one is merged to explore that.

**Review by pento — APPROVED** [review]
👍🏻 Lovely, than you for wrangling this, @dmsnell!
